sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/IcebergWriteSchemaTransformProvider.java

```diff
@@ -0,0 +1,183 @@
+    return Collections.singletonList(INPUT_TAG);
+  }
+
+  @Override
+  public List<String> outputCollectionNames() {
+    return Collections.singletonList(OUTPUT_TAG);
+  }
+
+  @Override
+  public String identifier() {
+    return ManagedTransformConstants.ICEBERG_WRITE;
+  }
+
+  @DefaultSchema(AutoValueSchema.class)
+  @AutoValue
+  public abstract static class Config {
+    public static Builder builder() {
+      return new AutoValue_IcebergWriteSchemaTransformProvider_Config.Builder();
+    }
+
+    @SchemaFieldDescription("Identifier of the Iceberg table to write to.")
+    public abstract String getTable();
+
+    @SchemaFieldDescription("Configuration parameters used to set up the Iceberg catalog.")
+    public abstract IcebergSchemaTransformCatalogConfig getCatalogConfig();
+
+    @AutoValue.Builder
+    public abstract static class Builder {
+      public abstract Builder setTable(String tables);
+
+      public abstract Builder setCatalogConfig(IcebergSchemaTransformCatalogConfig catalogConfig);
+
+      public abstract Config build();
+    }
+
+    public void validate() {
+      getCatalogConfig().validate();
+    }
+  }
+
+  static class IcebergWriteSchemaTransform extends SchemaTransform {
+    private final Config configuration;
+    private final Row configurationRow;
+
+    IcebergWriteSchemaTransform(Config configuration, Schema configSchema) {
+      this.configuration = configuration;
+
+      configurationRow =
+          Row.withSchema(configSchema)
+              .withFieldValue("table", configuration.getTable())
+              .withFieldValue("catalogConfig", configuration.getCatalogConfig().toRow())
+              .build();
+    }
+
+    Row getConfigurationRow() {
+      return configurationRow;
+    }
+
+    @Override
+    public PCollectionRowTuple expand(PCollectionRowTuple input) {
+
+      PCollection<Row> rows = input.get(INPUT_TAG);
+
+      IcebergSchemaTransformCatalogConfig catalogConfig = configuration.getCatalogConfig();
+
+      IcebergCatalogConfig.Builder catalogBuilder =
+          IcebergCatalogConfig.builder().setName(catalogConfig.getCatalogName());
+
+      if (!Strings.isNullOrEmpty(catalogConfig.getCatalogType())) {
+        catalogBuilder = catalogBuilder.setIcebergCatalogType(catalogConfig.getCatalogType());
+      }
+      if (!Strings.isNullOrEmpty(catalogConfig.getWarehouseLocation())) {
+        catalogBuilder = catalogBuilder.setWarehouseLocation(catalogConfig.getWarehouseLocation());
+      }
+
+      // TODO: support dynamic destinations
+      DynamicDestinations dynamicDestinations =
+          DynamicDestinations.singleTable(TableIdentifier.parse(configuration.getTable()));
+
+      IcebergWriteResult result =
+          rows.apply(IcebergIO.writeRows(catalogBuilder.build()).to(dynamicDestinations));
+
+      PCollection<Row> snapshots =
+          result
+              .getSnapshots()
+              .apply(MapElements.via(new SnapshotToRow()))
```

Comment:
FYI, it may not be enough to just map to Rows, as all the intermediate data choices will be sticky for pipeline update. (We may be stuck with this for some existing sources, but we should think long and hard about what this means for anything new.)